### PR TITLE
[HACK week] Product list - Add Favorites section 

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Part of: #12274

## Description
List favorite products under a new section in product list table view. 

Changes
- Limit the max number of products that can be marked as favorite to avoid taking too much space in product list screen. In future, we can consider making "Favorites" section collapsible to make space for the "All" section.
- Load the favorites product IDs from remote and show the products under "Favorites" section.
- Observe to user defaults and reload when a product is marked/removed as favorite. 

## Testing instructions
- Login into the app
- Navigate to Products tab
- Open any product
- Tap "…" from the action sheet tap "Mark as favorite"
- Navigate back to the product list screen.
- You should see the product under "Favorites" section
- Open the same product again
- Tap "…" -> "Remove from a favourite"  on it to remove the product from the favourite list.
- Navigate back to the product list screen and validate that the "Favorites" section is updated.
- Mark a few products to the favorites section.
- Switch stores and ensure that the products marked as favorite are still marked as favorite. 
- Log out and log into the same store again. Validate that the favorite products list is cleared after logout.

## Screenshots

Simulator Screen Recording - iPhone 15 Pro - 2024-03-15 at 20 16 26

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Products/FavoriteProducts/FavoriteProductsUseCase.swift [hunk 1]
```diff
@@ -21,7 +21,7 @@ struct FavoriteProductsUseCase {
     func markAsFavorite(productID: Int64) {
         if var productIdDict = productIdDict {
             let existingFavProductIDs = productIdDict[idAsString] ?? []
-            productIdDict[idAsString] = Array(Set(existingFavProductIDs + [productID]))
+            productIdDict[idAsString] = Array(Set(existingFavProductIDs + [productID])).suffix(Constants.favoriteProductsMaxLimit)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/FavoriteProducts/FavoriteProductsUseCase.swift]: Setting this limit would remove the older items from the favorite list right? I don't think this is easy for users to understand, they might question where the other items go. Should we keep the list under the feature flag and add the view more option when we have time?

File: WooCommerce/Classes/ViewRelated/Products/FavoriteProducts/ProductListViewSection.swift [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import Foundation
+import struct Yosemite.Product
+
+struct ProductListViewSection {
+
+    enum ProductListViewSectionType {
+        case favourites
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/FavoriteProducts/ProductListViewSection.swift]: Super nit: let's keep `favorite` naming for consistency across different places.

File: WooCommerce/Classes/ViewRelated/Products/ProductsListViewModel.swift [hunk 3]
```diff
@@ -23,6 +23,7 @@ final class ProductListViewModel: ProductsListViewModelProtocol {
 
     private let barcodeSKUScannerItemFinder: BarcodeSKUScannerItemFinder
     private let featureFlagService: FeatureFlagService
+    let favoriteProductsUseCase: FavoriteProductsUseCase
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/ProductsListViewModel.swift]: Super nit: do we want to keep this private as well?

File: WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift [hunk 4]
```diff
@@ -1012,26 +1035,61 @@ private extension ProductsViewController {
     }
 }
 
+// MARK: - Favorite products
+//
+private extension ProductsViewController {
+    func observeFavoriteProductIDChanges() {
+        userDefaults.publisher(for: \.favoriteProductIDs)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift]: Nit: we can create a publisher from the use case and access it through the view model without revealing the implementation detail with user defaults here.

File: WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift [hunk 5]
```diff
@@ -1562,6 +1641,32 @@ extension ProductsViewController {
...
+        if favoriteProducts.isNotEmpty {
+            sections = [ProductListViewSection(type: .favourites, products: favoriteProducts),
+                        ProductListViewSection(type: .allProducts, products: resultsController.fetchedObjects)]
+        } else {
+            sections = [ProductListViewSection(type: .allProducts, products: resultsController.fetchedObjects)]
+        }
+    }
+
+    func indexPath(for productID: Int64) -> IndexPath? {
+        guard let allProductsSectionIndex = sections.firstIndex(where: {$0.type == .allProducts }),
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift:1708]: Nit: Can you leave a comment explaining why we're supporting index path only for the All Products section?

**Review by itsmeichigo — COMMENTED** [review]
This works as described in the PR description 👍 I have a few concerns:
- Regarding the logic of the previous PR, I wonder if we should switch to saving favorite products in a local file with `GeneralAppSettings` rather than user defaults. IMO we should keep only simpler checks in user defaults, especially those that require synchronous requests. Favorite products might fit better with local file storage.
- I'm unsure about the limit of the favorite items, as it's unclear to users when their older items are removed. We should keep all items and if that's too clustering, we can keep the feature under a feature flag before adding the view more option.

**Comment by selanthiraiyan** [comment]
Thanks for the review, Huong! I am marking the PR as a draft until I get time to address the feedback.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Products/ProductsListViewModel.swift (hunk 3 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/ProductsListViewModel.swift]: Done in https://github.com/woocommerce/woocommerce-ios/pull/12287/commits/9b9e11aed86de22a62ceb5e2403c455c816066fe

File: WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift (hunk 4 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift]: Now that we have started using AppSettings to store the favorite product IDs, I have added a callback to listen to mark/remove favorite changes.

File: WooCommerce/Classes/ViewRelated/Products/FavoriteProducts/ProductListViewSection.swift (hunk 2 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/FavoriteProducts/ProductListViewSection.swift]: Done in https://github.com/woocommerce/woocommerce-ios/pull/12287/commits/836c3c48e269d1901196bf8d526dfb154e5d9e24

File: WooCommerce/Classes/ViewRelated/Products/FavoriteProducts/FavoriteProductsUseCase.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/FavoriteProducts/FavoriteProductsUseCase.swift]: I have removed this limit now as we added a way to expand/collapse the favorites section.

File: WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift (hunk 5 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift:1708]: I have renamed the method in https://github.com/woocommerce/woocommerce-ios/pull/12287/commits/d90870480206b507c732fec59fc00649d77a9c35

This is used to select the product when changes are made in a split view. I don't understand the nuances of the split view part yet. I would prefer to keep the selected product and scroll position untouched until we get time to revisit this work.

File: WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift [hunk 6]
```diff
@@ -1416,30 +1538,32 @@ extension ProductsViewController: PaginationTrackerDelegate {
...
     ///
-    private func syncLocalProductsSettings(onCompletion: @escaping (Result<StoredProductSettings.Setting, Error>) -> Void) {
-        let action = AppSettingsAction.loadProductsSettings(siteID: siteID) { [weak self] (result) in
-            switch result {
-            case .success(let settings):
-                self?.syncProductCategoryFilterRemotely(from: settings) { [weak self] settings in
-                    guard let self else { return }
-                    if let sort = settings.sort {
-                        sortOrder = ProductsSortOrder(rawValue: sort) ?? .default
+    private func syncLocalProductsSettings() async throws -> StoredProductSettings.Setting {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift:1541]: Made this method `async` and the rest are just indentation changes.

**Comment by selanthiraiyan** [comment]
👋 @itsmeichigo , Thank you for the review and input. 🙇 

I have pushed the necessary changes to make the favorites section collapse/expand. The PR is ready for another round of review. 


Please note that I don't check for feature flag anywhere because,
- The favorites section will be displayed only if there are favorite products.
- The action buttons are enabled only if the feature flag is turned on. https://github.com/woocommerce/woocommerce-ios/pull/12284 This means that users can mark a product favorite only if the feature flag is turned on.
- Checking feature flag inside `ProductsViewController` further complicates the code and might create edge case issues. 


#### Changes include
- Two sections in products table view. "Favorites" and "All"
- "Favorites" section can be collapsed and expanded by the user.
- Custom section header view with a button to collapse/expand section.
- Remove background from `PlainTextSectionHeaderView` for using this in products table view. This isn't used anywhere else.
- Callback when a product is marked or removed as favorite. We reload the tableview from this callback to update favorites section.

#### Known scrolling issue
When you select a product from the "Favorites" section and navigate back, the product will be selected in the "All" section
- The selection and scrolling is added as part of the split view controller work. https://github.com/woocommerce/woocommerce-ios/pull/12060 I don't fully understand the reasons, though. 
- I preferred to keep that untouched to avoid creating undesirable effects. (Product creation, editing, deletion, split view etc)
- I am tracking this issue at https://github.com/woocommerce/woocommerce-ios/issues/12274, and we can revisit it when we find the time. (Maybe next HACK week) 

@jaclync 👋  Kindly share if you have any ideas or input about this issue. Thank you!

#### Demo

**iPhone**

https://github.com/woocommerce/woocommerce-ios/assets/524475/e083326a-daab-4361-b68b-48eebed56c26


**iPad**

https://github.com/woocommerce/woocommerce-ios/assets/524475/80faa643-da6b-4e60-97e8-3873503131a5

**Review by itsmeichigo — COMMENTED** [review]
This works as described (again). I appreciate your efforts, but I have some suggestions for improvements:

- Having the favorite list as a section comes with a few complications like the selection state you mentioned. I also see that when selecting a filter, showing the favorite section can be misleading as the section is not filtered.
- Some minor UI issues: the product list has a divider at the top, while there's no divider between the Favorites and All sections, which looks odd. There should be a divider between the sections for better clarity:

 

For a simpler UX, we can remove the Favorite section and update the product rows with an additional star icon for favorite items instead. Then, we can add an option to see only favorite items in the Filter list.

**Comment by selanthiraiyan** [comment]
Thanks again for the review, Huong!

> For a simpler UX, we can remove the Favorite section and update the product rows with an additional star icon for favorite items instead. Then, we can add an option to see only favorite items in the Filter list.

This is a good suggestion. I will try to address this when I find time. I will mark the PR as a draft until then.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by selanthiraiyan** [comment]
Closing this PR. I will open other PRs to implement the favorite product filter.

**Pull request closed without merging**